fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+        messages: vec![],
+        log,
+        data: None,
+    };
+    Ok(r)
+}
+
+fn try_reverse_swap<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: &Env,
+    state: &State,
+    rid: u64,
+    to: HumanAddr,
+    from: String,
+    origin_tx_hash: Uint128,
+    amount: Uint128,
+    relay_eon: u64,
+) -> StdResult<HandleResponse> {
+    verify_tx_relay_eon(relay_eon, state)?;
+    only_relayer(env, state)?;
+
+    if amount > state.swap_fee {
+        // TOFIX(LR) when amount == fee, amount will still be consumed
+        let swap_fee = state.swap_fee;
+        let effective_amount = (amount - swap_fee)?;
+        let to_canonical = deps.api.canonical_address(&to)?;
+        let rtx = send_tokens(
+            &deps.api,
+            &env,
+            &state,
+            &to_canonical,
+            amount,
+            "reverse_swap",
+        )?;
+        config(&mut deps.storage).update(|mut state| {
+            state.supply = (state.supply - amount)?;
+            state.sealed_reverse_swap_id += 1; // TOFIX(LR) after exec should be == rid
+                                               //state.sealed_reverse_swap_id = rid;
+            Ok(state)
+        })?;
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: rtx.messages,
+            log,
+            data: None,
+        };
+        Ok(r)
+    } else {
+        let swap_fee = amount;
+        let effective_amount = Uint128::zero();
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: vec![],
+            log,
+            data: None,
+        };
+        Ok(r)
+    }
+}
+
+/* ***************************************************
+ * *****************    Helpers      *****************
+ * ***************************************************/
+
+fn amount_from_funds(funds: &Vec<Coin>) -> Uint128 {
+    // TODO(LR) return proper error
+    if funds.len() == 1
+    /* && funds[0].denom == "fet" */
+    {
+        // TODO(LR) does cosmwas allows sending multiple funds of the same token
+        funds[0].amount
+    } else {
+        Uint128::zero()
+    }
+}
+
+fn send_tokens<A: Api>(
+    api: &A,
+    env: &Env,
+    state: &State,
+    to_address: &CanonicalAddr,
+    amount: Uint128,
+    action: &str,
+) -> HandleResult {
+    let from_address = &env.contract.address;
+    let from_human = api.human_address(&from_address)?;
```

Comment:
I would sugges to rather use `contract_addr` & `contract_addr_human` than `from_address` & `from_human`, since it will be clear to read what the variable represents.

Also, in order to avoid calling `api.human_address(...)` conversion **each** time for the very same value (contract address, which is conctant for whole contract life-cycle), I would suggest to do this:
1. check whether `env.contract` object does not already contain property for "human" formatted contract address
2. if above fails, then simply call this conversion only oince in the contract contructor and store it in the state - try to find out whether it is possible to create **immutable** contract state = it can be set just **once** in the contructror

Reply:
Btw, it is appealing that we need to call do this address dencoding in the first palce - first that Send message can take on board only "human" encoded address, and/or that contract addres is not **already** offered also in "human" encoding in contract context ...